File: Sources/Network/_HttpDatasource.m
//
//  Example
//  man
//
//  Created by man 11/11/2018.
//  Copyright © 2020 man. All rights reserved.
//

#import "_HttpDatasource.h"
#import "_NetworkHelper.h"

static const NSUInteger kDefaultMemoryBudget = 50 * 1024 * 1024; // 50 MB

@interface _HttpDatasource ()
@property (nonatomic, assign, readwrite) NSUInteger totalDataSize;
@end

@implementation _HttpDatasource

+ (instancetype)shared
{
    static id sharedInstance = nil;

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });

    return sharedInstance;
}

- (id)init
{
    self = [super init];
    if (self) {
        self.httpModels = [NSMutableArray arrayWithCapacity:1000 + 100];
        self.totalDataSize = 0;
        // Clear leftover disk cache from previous app session
        [_HttpModel clearDiskCache];
    }
    return self;
}

- (NSUInteger)estimatedSizeOfModel:(_HttpModel *)model
{
    // Use cached sizes - data is on disk, don't load it just to check length
    NSUInteger size = 0;
    size += model.requestDataSize;
    size += model.responseDataSize;
    size += 1024; // overhead for URL, headers, strings, etc.
    return size;
}

- (BOOL)addHttpRequset:(_HttpModel*)model
{
    if ([model.url.absoluteString isEqualToString:@""]) {
        return NO;
    }

    //url Filter, ignore case
    for (NSString *urlString in [[_NetworkHelper shared] ignoredURLs]) {
        if ([[model.url.absoluteString lowercaseString] containsString:[urlString lowercaseString]]) {
            return NO;
        }
    }

    // All mutations to httpModels must be synchronized - stopLoading is called
    // from different protocol instance threads concurrently.
    @synchronized (self) {
        //Maximum number limit
        if (self.httpModels.count >= 1000) {
            if ([self.httpModels count] > 0) {
                _HttpModel *oldest = self.httpModels[0];
                self.totalDataSize -= [self estimatedSizeOfModel:oldest];
                [self.httpModels removeObjectAtIndex:0];
            }
        }

        //detect repeated
        __block BOOL isExist = NO;
        [self.httpModels enumerateObjectsUsingBlock:^(_HttpModel *obj, NSUInteger index, BOOL *stop) {
            if ([obj.requestId isEqualToString:model.requestId]) {
                isExist = YES;
                *stop = YES;
            }
        }];
        if (isExist) {
            return NO;
        }

        // Enforce disk budget: evict oldest models until we have room
        NSUInteger modelSize = [self estimatedSizeOfModel:model];
        while (self.totalDataSize + modelSize > kDefaultMemoryBudget && self.httpModels.count > 0) {
            _HttpModel *oldest = self.httpModels[0];
            self.totalDataSize -= [self estimatedSizeOfModel:oldest];
            [self.httpModels removeObjectAtIndex:0];
            // Note: model's dealloc deletes its disk files automatically
        }

        [self.httpModels addObject:model];
        self.totalDataSize += modelSize;
    }

    return YES;
}

- (void)reset
{
    @synchronized (self) {
        [self.httpModels removeAllObjects];
        self.totalDataSize = 0;
        // Clear all disk files
        [_HttpModel clearDiskCache];
    }
}

- (void)remove:(_HttpModel *)model
{
    @synchronized (self) {
        [self.httpModels enumerateObjectsWithOptions:NSEnumerationReverse usingBlock:^(_HttpModel *obj, NSUInteger index, BOOL *stop) {
            if ([obj.requestId isEqualToString:model.requestId]) {
                self.totalDataSize -= [self estimatedSizeOfModel:obj];
                [self.httpModels removeObjectAtIndex:index];
                // Note: model's dealloc deletes its disk files automatically
            }
        }];
    }
}

@end

File: Sources/Network/HttpDatasource.swift
//
//  Example
//  man
//
//  Created by man 11/11/2018.
//  Copyright © 2020 man. All rights reserved.
//

import Foundation

private let kDefaultMemoryBudget: UInt = 50 * 1024 * 1024 // 50 MB

@objc class _HttpDatasource: NSObject {

    @objc var httpModels: NSMutableArray
    @objc private(set) var totalDataSize: UInt = 0

    @objc static let shared = _HttpDatasource()

    private override init() {
        httpModels = NSMutableArray(capacity: 1000 + 100)
        totalDataSize = 0
        super.init()
        // Clear leftover disk cache from previous app session
        _HttpModel.clearDiskCache()
    }

    private func estimatedSizeOfModel(_ model: _HttpModel) -> UInt {
        // Use cached sizes - data is on disk, don't load it just to check length
        var size: UInt = 0
        size += UInt(model.requestDataSize)
        size += UInt(model.responseDataSize)
        size += 1024 // overhead for URL, headers, strings, etc.
        return size
    }

    /// NOTE: Keeping the typo "Requset" in the method name for compatibility
    @objc func addHttpRequset(_ model: _HttpModel) -> Bool {
        if model.url?.absoluteString == "" {
            return false
        }

        // url Filter, ignore case
        for urlString in _NetworkHelper.shared.ignoredURLs ?? [] {
            if model.url?.absoluteString?.lowercased().contains(urlString.lowercased()) ?? false {
                return false
            }
        }

        // All mutations to httpModels must be synchronized - stopLoading is called
        // from different protocol instance threads concurrently.
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        // Maximum number limit
        if httpModels.count >= 1000 {
            if httpModels.count > 0 {
                if let oldest = httpModels[0] as? _HttpModel {
                    let sizeToRemove = estimatedSizeOfModel(oldest)
                    totalDataSize = totalDataSize > sizeToRemove ? totalDataSize - sizeToRemove : 0
                }
                httpModels.removeObject(at: 0)
            }
        }

        // detect repeated (guard against nil — in ObjC [nil isEqualToString:nil] returns NO)
        var isExist = false
        for i in 0..<httpModels.count {
            if let obj = httpModels[i] as? _HttpModel {
                if let rid = obj.requestId, let mrid = model.requestId, rid == mrid {
                    isExist = true
                    break
                }
            }
        }
        if isExist {
            return false
        }

        // Enforce disk budget: evict oldest models until we have room
        let modelSize = estimatedSizeOfModel(model)
        while totalDataSize + modelSize > kDefaultMemoryBudget && httpModels.count > 0 {
            if let oldest = httpModels[0] as? _HttpModel {
                let sizeToRemove = estimatedSizeOfModel(oldest)
                totalDataSize = totalDataSize > sizeToRemove ? totalDataSize - sizeToRemove : 0
            }
            httpModels.removeObject(at: 0)
            // Note: model's dealloc deletes its disk files automatically
        }

        httpModels.add(model)
        totalDataSize += modelSize

        return true
    }

    @objc func reset() {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        httpModels.removeAllObjects()
        totalDataSize = 0
        // Clear all disk files
        _HttpModel.clearDiskCache()
    }

    @objc func remove(_ model: _HttpModel) {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        // Enumerate in reverse to safely remove while iterating
        for i in stride(from: httpModels.count - 1, through: 0, by: -1) {
            if let obj = httpModels[i] as? _HttpModel {
                if let rid = obj.requestId, let mrid = model.requestId, rid == mrid {
                    let sizeToRemove = estimatedSizeOfModel(obj)
                    totalDataSize = totalDataSize > sizeToRemove ? totalDataSize - sizeToRemove : 0
                    httpModels.removeObject(at: i)
                    // Note: model's dealloc deletes its disk files automatically
                }
            }
        }
    }
}
